azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt

```diff
@@ -0,0 +1,105 @@
+package com.azure.android.communication.ui.calling.presentation.manager
+
+import android.content.Context
+import android.os.Build
+import android.os.Handler
+import android.os.Looper
+import android.telephony.PhoneStateListener
+import android.telephony.TelephonyCallback
+import android.telephony.TelephonyManager
+import androidx.annotation.RequiresApi
+
+import com.azure.android.communication.ui.calling.redux.Store
+import com.azure.android.communication.ui.calling.redux.action.CallingAction
+import com.azure.android.communication.ui.calling.redux.state.CallingStatus
+import com.azure.android.communication.ui.calling.redux.state.PermissionStatus
+import com.azure.android.communication.ui.calling.redux.state.ReduxState
+import kotlinx.coroutines.flow.collect
+import java.util.concurrent.Executor
+
+// Detects phone calls and hangs up if accepted.
+internal class DetectPhoneCallManager(val context: Context, val store: Store<ReduxState>) {
+    private var detectPhoneCall: DetectPhoneCall? = null
+
+    suspend fun onCreate() {
+        store.getStateFlow().collect {
+            if (it.permissionState.phonePermissionState == PermissionStatus.GRANTED) {
+                detectPhoneCall = DetectPhoneCall.register(context, store)
+            }
+        }
+    }
+
+    fun onDestroy() = detectPhoneCall?.unregister()
+}
+
+// / Common Interface for the API >= 31 and < 31 mechanisms to do this
+private interface DetectPhoneCall {
+    companion object {
+        fun register(context: Context, store: Store<ReduxState>): DetectPhoneCall {
+            val tm = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
+            var receiver: DetectPhoneCall = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                DetectPhoneCallV31(tm, store)
+            } else {
+                DetectPhoneCallLegacy(tm, store)
+            }
+            receiver.register()
+            return receiver
+        }
+    }
+
+    val store: Store<ReduxState>
+    fun register()
+    fun unregister()
+    fun onOffHook() {
+        if (store.getCurrentState().callState.callingStatus == CallingStatus.CONNECTED) {
+            store.dispatch(CallingAction.CallEndRequested())
+        }
+    }
+}
+
+// / Legacy Version for < 31
+private class DetectPhoneCallLegacy(private val telephonyManager: TelephonyManager, override val store: Store<ReduxState>) :
+    PhoneStateListener(), DetectPhoneCall {
```

Comment:
`PhoneStateListener` is deprecated, can we add `@Suppress("DEPRECATION")`

Reply:
Done